### PR TITLE
 Bug 1927168: Fail fast when applying resources with no definition

### PR DESCRIPTION
**Pull request by arjunrn**
The CVO attempts to retry creation of resources when there are failures. But this approach prevents CVO from advancing when the resource defition for a resource is missing. The CVO should instead no rety and proceed with the creation of subsequent resources so that the definition could be applied later.

### CODE REVIEW

File: pkg/cvo/sync_test.go [hunk 1]
```diff
@@ -41,74 +41,77 @@ func Test_SyncWorker_apply(t *testing.T) {
 
 		check   func(*testing.T, []action)
 		wantErr bool
-	}{{
```
Comment (arjunrn) [inline pkg/cvo/sync_test.go:44]: Note to Reviewers: Although it looks like a lot of this test has changed, the only non-formatting change is the change in number of `Actions`.

Comment (wking) [reply pkg/cvo/sync_test.go:44]: easy way to help review with that sort of thing is to isolate the no-op format changes in a separate commit, so folks looking at the full PR can focus on the semantic changes.

Comment (arjunrn) [reply pkg/cvo/sync_test.go:44]: Moved it into a separate commit

File: pkg/cvo/internal/generic_test.go [hunk 2]
```diff
@@ -87,3 +87,71 @@ func TestCreateOnlyUpdate(t *testing.T) {
...
+}`,
+		},
+		{
+			name: "delete annotation",
+			resource: `{
+  "kind": "UnknownType",
+  "apiVersion": "config.openshift.io/v1",
+  "metadata": {
+     "name": "cluster",
+	 "annotations": {
```
Comment (wking) [inline pkg/cvo/internal/generic_test.go:113]: nit: seems like you're adding an additional level of indentation here?

File: pkg/cvo/internal/generic_test.go [hunk 3]
```diff
@@ -87,3 +87,71 @@ func TestCreateOnlyUpdate(t *testing.T) {
...
+		 "release.openshift.io/delete": "true"
+	 }
+  }
+}`,
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			ctx := context.Background()
+			obj, err := runtime.Decode(unstructured.UnstructuredJSONScheme, []byte(tc.resource))
```
Comment (wking) [inline pkg/cvo/internal/generic_test.go:122]: you can probably avoid some of these runtime gymnastics by including `github.com/openshift/library-go/pkg/manifest` and using `json.Unmarshal` like we do [here][1].

[1]: https://github.com/openshift/cluster-version-operator/blob/887aa47093d740ce827e3046848d4f0c1aca374f/pkg/cvo/sync_test.go#L116-L118

File: pkg/cvo/internal/generic_test.go (hunk 2 above)
Comment (arjunrn) [reply pkg/cvo/internal/generic_test.go]: Mixing tabs and spaces got me there. Fixed now.

File: pkg/cvo/internal/generic_test.go (hunk 3 above)
Comment (arjunrn) [reply pkg/cvo/internal/generic_test.go]: Updated as per recommendation.

File: pkg/cvo/internal/generic.go [hunk 4]
```diff
@@ -42,6 +43,12 @@ func deleteUnstructured(ctx context.Context, client dynamic.ResourceInterface, r
 		Name:      required.GetName(),
 	}
 	existing, err := client.Get(ctx, required.GetName(), metav1.GetOptions{})
+
+	// If there is a NoMatchError then, it's guaranteed that the resource does not exist.
+	if err != nil && meta.IsNoMatchError(err) {
+		return false, err
```
Comment (wking) [inline pkg/cvo/internal/generic.go:49]: hmm, but if this is a delete manifest, this isn't an error, this is "hooray, success", right?  Maybe `GetDeleteProgress` needs to be taught to distinguish this case?

Comment (jottofar) [reply pkg/cvo/internal/generic.go:49]: Agree, it should be handled in `GetDeleteProgress` which is intended to handle all returns from the `Get`. [Here](https://github.com/openshift/cluster-version-operator/blob/3a68652568e9075c23f491bc8c037942bd67ec82) and [here](https://github.com/openshift/cluster-version-operator/blob/3a68652568e9075c23f491bc8c037942bd67ec82/lib/resourcedelete/helper.go#L124) as :
```
if apierrors.IsNotFound(getError) || meta.IsNoMatchError(getError) {
```

Comment (arjunrn) [reply pkg/cvo/internal/generic.go]: Updated to handle it in `GetDeleteProgress`.

File: pkg/cvo/internal/generic_test.go [hunk 5]
```diff
@@ -87,3 +88,74 @@ func TestCreateOnlyUpdate(t *testing.T) {
...
+			}
+			err = builder.Do(ctx)
+			if tc.expectErr {
+				if err == nil {
+					t.Fatal("expected error, got none")
+				}
+				if !meta.IsNoMatchError(err) {
+					t.Fatalf("expected error of type IsNoMatchError, instead got: %T", err)
+				}
+			}
```
Comment (wking) [inline pkg/cvo/internal/generic_test.go:158]: To handle the "we didn't expect an error, but got one" case:

```go
} else if err != nil {
  t.Fatal(err)
}
```

**Pull request closed without merging**